Source/WebCore/page/LocalDOMWindow.cpp

```diff
@@ -2458,39 +2460,188 @@ void LocalDOMWindow::finishedLoading()
     }
 }
 
-PerformanceEventTiming::Candidate LocalDOMWindow::initializeEventTimingEntry(const Event& event, EventTypeInfo typeInfo)
+InteractionID LocalDOMWindow::computeInteractionID(const Event& event, EventType type)
 {
-    // FIXME: implement InteractionId logic
+    switch (type) {
+    case EventType::keyup: {
+        ASSERT(event.isKeyboardEvent());
+        auto keyboardEvent = downcast<KeyboardEvent>(&event);
+        ASSERT(keyboardEvent);
+        if (keyboardEvent->isComposing())
+            return 0;
+
+        auto it = m_pendingKeyDowns.find(keyboardEvent->keyCode());
+        if (it == m_pendingKeyDowns.end())
+            return 0;
+
+        auto ret = increaseInteractionCount();
+        it->value.interactionID = ret;
+        m_performanceEventTimingCandidates.append(it->value);
+        m_pendingKeyDowns.remove(it);
+        return ret;
+    }
+    case EventType::compositionstart: {
+        for (auto& pendingEntry : m_pendingKeyDowns)
+            m_performanceEventTimingCandidates.append(pendingEntry.value);
+
+        m_pendingKeyDowns.clear();
+        return 0;
+    }
+    case EventType::input: {
+        if (!event.isInputEvent())
+            return 0;
+
+        ASSERT(event.isInputEvent());
+        auto inputEvent = downcast<InputEvent>(&event);
+        ASSERT(inputEvent);
+        if (!inputEvent->isInputMethodComposing())
+            return 0;
+
+        return increaseInteractionCount();
+    }
+    case EventType::click: {
+        if (!m_pointerInteractionState.interactionID)
```

Comment:
Why do we not have `pointerMap`? We should at least clarify why in the commit message.